src/runtime/tuple.c

```diff
@@ -143,8 +143,18 @@ value decode_value(heap h, tuple dictionary, buffer source)
                 s = table_find(dictionary, pointer_from_u64(nlen));
                 if (!s) halt("indirect symbol not found: 0x%lx, offset %d\n", nlen, source->start);
             }
-            value nv = decode_value(h, dictionary, source);
+            value nv = decode_value(h, dictionary, source, total, obsolete);
+            if (obsolete) {
+                value old_v = table_find(t, s);
+                if (old_v) {
+                    (*obsolete)++;
+                    if (!nv)
+                        (*obsolete)++;
```

Comment:
If the value of a tuple is being deleted at a given point, it means there are (at least) 2 unnecessary log entries: the one that sets the initial value, and the one that removes the value. So if both entries were removed, the log would still be consistent with the current state. So "obsolete" is meant to count the number of log entries that would not be there if the log were to be recreated "cleanly". I know "obsolete" is not exactly the right word here, because in the above scenario the latter log entry is needed to undo what the former entry did, but I couldn't come up with a better naming.